solr/src/main/java/dk/dbc/opensearch/solr/flatquery/FlatQuery.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (C) 2018 DBC A/S (http://dbc.dk/)
+ *
+ * This is part of opensearch-solr
+ *
+ * opensearch-solr is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * opensearch-solr is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a modifiableCopy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package dk.dbc.opensearch.solr.flatquery;
+
+import dk.dbc.opensearch.cql.BoolQuery;
+import dk.dbc.opensearch.cql.QueryNode;
+import dk.dbc.opensearch.cql.SearchQuery;
+import dk.dbc.opensearch.cql.token.BooleanOpName;
+import dk.dbc.opensearch.solr.config.FieldSpec;
+import dk.dbc.opensearch.solr.SolrRules;
+
+/**
+ * Abstract class representing flattened query.
+ *
+ * @author DBC {@literal <dbc.dk>}
+ */
+public abstract class FlatQuery {
+
+    /**
+     * Map a Query Tree to a Flat tree given a set of rules
+     *
+     * @param solrRules the index rules
+     * @param query     the query tree
+     * @return a flat query representation
+     */
+    public static FlatQuery from(SolrRules solrRules, QueryNode query) {
+        FlatQuery flat = new Flattener(solrRules).flatten(query);
+        if (flat instanceof FlatQuerySearch) {
+            flat = new FlatQueryAndNot(flat);
+        }
+
+        return flat;
+    }
+
+    /**
+     * Tell if all nested queries can be put into same nested query
+     *
+     * @param spec the nested query spec wanted
+     * @return true/false
+     */
+    public abstract boolean allAreNestedOfType(FieldSpec spec);
+
+    /**
+     * Tell if all nested queries can be put into same nested query
+     *
+     * @param spec the nested query spec wanted
+     * @return true/false
+     */
+    public abstract boolean anyAreNestedOfType(FieldSpec spec);
+
+    /**
+     * Tell if all subtree nodes can be put into a filterquery
+     *
+     * @param filterQueryName
+     * @return true/false
+     */
+    public abstract boolean allAreFilterQuery(String filterQueryName);
+
+    private static class Flattener {
+
+        private final SolrRules solrRules;
+
+        Flattener(SolrRules solrRules) {
+            this.solrRules = solrRules;
```

Comment:
Er der brug for en `Flattener` klasse? Den har jo ingen interne fields. Kunne `flatten`, `flattenSearch`, `flattenNot` og `flattenOr` ikke være private metoder på `FlatQuery`?

Reply:
Jeg syntes solrRules er er internt felt, som ikke er relevant på FlatQuery klassen.